instrumentation/zio/zio-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/zio/v2_0/FiberContext.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.zio.v2_0;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+
+public class FiberContext {
+
+  private Span span;
+
+  private FiberContext(Span span) {
+    this.span = span;
+  }
+
+  public static FiberContext create() {
+    return new FiberContext(Span.current());
+  }
+
+  public void onEnd() {
+    Context.root().makeCurrent();
+  }
+
+  public void onSuspend() {
+    this.span = Span.current();
+    Context.root().makeCurrent();
```

Comment:
Every time you call `makeCurrent` on `Context` you should also call `close` on the returned `Scope`. Failing to do so may result in leaked scopes that make the test fail. For example see https://github.com/open-telemetry/opentelemetry-java/blob/main/extensions/kotlin/src/main/java/io/opentelemetry/extension/kotlin/KotlinContextElement.java
Instead of tracking current `Span` it would be better to use current `Context`.

Reply:
Hi @laurit, thank you for taking a look at this PR!

As I wrote in "Notes about implementation", I've tried multiple versions, including with `Context.makeCurrent()`, similarly to what's done in `KotlinContextElement`. And I've tried to track and close scopes as well. None of these versions worked. The only version that reliably works is this one, without attempts to close scopes.

Sure, this implementation does leak scopes. However in default configuration with `ThreadLocalContextStorage` scopes are just references to the previous context, which is re-activated when scope is closed. Not closing scope doesn't seem to have any consequences in this case.

I guess it's not 100% correct in general case as some other implementations of `ContextStorage` might rely on scopes being closed. But I'd say that having instrumentation in most cases is way better than not having it at all.

Reply:
I've replaced `Span.makeCurrent` with `Context.makeCurrent`. It works, but only if I take a snapshot of the current context in `onSuspend` (https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/7980/commits/a41f879ba96e2d19f6d29f9f71e368713cc2ce5e.

Reply:
How can I reproduce the issue you are seeing when you close the scope? It would be nice if we could figure out why it isn't working as it is supposed to work.

Reply:
My understanding is that in `onSuspend` you should have `this.context = Context.current()`. You need to remember the context at the time of suspension so you could restore it when the fiber resumes.
For `ThreadLocalContextStorage` log message there are 2 reasons. Firstly sometimes `onEnd` and `onSuspend` are called in sequence which will result in a double close. Secondly this seems to happen when scopes are closed out of order. I'll take this up with opentelemetry-java maintainers and ask their guidance.

Reply:
> My understanding is that in `onSuspend` you should have `this.context = Context.current()`. You need to remember the context at the time of suspension so you could restore it when the fiber resumes.

Yes, I've realized that `Context` is immutable, so we definitely need to do this.

> Secondly this seems to happen when scopes are closed out of order. 

I've debugged the sequence of events and it seems that the following is happening:

1. While fiber is running, it opens a new span.
2. Fiber is suspended.
3. Fiber is resumed, so we create a new scope.
4. Fiber ends the current span.
5. Fiber is suspended and current context is not the same as it was in step 3. So we are trying to close span for context that's not there anymore.

I guess this is what you meant when you wrote that "scopes are closed out of order"?

Reply:
> Firstly sometimes onEnd and onSuspend are called in sequence.

I think this happens each time when fiber is being terminated. Both `onEnd` and `onSuspend` are called. Even if fiber was interrupted. Since `onSuspend` is always called when fiber is terminating, I've removed the `onEnd` callback.

Reply:
Since contexts are immutable, it seems that closing scope in `onSuspend` more often then not is a noop. Because if fiber started a child span, for example, then current context would be different from what `scope.close()` would expect.

And even if we manage to close all open scopes somehow, we would be still leaking one scope per fiber anyway - the one created by last call to `Context.root().makeCurrent()` in `onSuspend` , when fiber is being terminated.

So I'm wondering - what if we go with the original version that didn't attempt to close scopes at all?

Reply:
@laurit thank you, merged.